hilti/toolchain/src/compiler/codegen/ctors.cc

```diff
@@ -181,7 +181,21 @@ struct Visitor : hilti::visitor::PreOrder<cxx::Expression, Visitor> {
         return fmt("::hilti::rt::Stream(\"%s\"_b)", util::escapeBytesForCxx(n.value()));
     }
 
-    result_t operator()(const ctor::String& n) { return fmt("std::string(\"%s\")", util::escapeUTF8(n.value(), true)); }
+    result_t operator()(const ctor::String& n) {
+        // We emit string constructors for both user-provided literals and
+        // generated strings, e.g., static error messages or identifiers. Only
+        // user-provided literals need to behave like actual our runtime string
+        // type `std::string` since they e.g., need to be usable with functions
+        // expecting that type or might need to be mutable; generated string
+        // OTOH are static and only used with our internal APIs which we can
+        // make usable with actual literals, and we can avoid allocating owning
+        // mutable strings for them.
```

Comment:
@rsmmr This is a lot of magic here, not sure you feel about this. The upside is that we can keep using `builder::string` without having to think too much about what we stringify. The alternative could be to e.g., introduce a separate builder for generated literals.

Reply:
Yeah, not feeling very good about this honestly. That'd be very hard to remember and track down if anything goes wrong either way. Either a separate builder or make it explicit somehow in the meta data that's not from a user (i.e., not just non-existence of location). I think I'd indeed prefer the former, as you already suggest.

Reply:
I cleaned this up somewhat to track in a `ctor::String` whether it was a literal or we'd need to allocate. With what I have now we'd need to make a deliberate choice between `builder::string` and `builder::string_literal`. As we move more internal methods over to accept `string_view` incorrectly using `builder::string` with generated literals would not compile anymore (this is good), but we could still accidentally slip in `builder::string` instead of `builder::string_literal` so we'd do more work.

Reply:
I went with `builder::string_mut`. I also added a new `builder::string` which is immediately marked deprecated, but forwards to `builder::string_mut`.